repo/migrations/Migration012.go

```diff
@@ -0,0 +1,257 @@
+package migrations
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/OpenBazaar/jsonpb"
+	"github.com/OpenBazaar/openbazaar-go/ipfs"
+	"github.com/OpenBazaar/openbazaar-go/pb"
+	"github.com/gogo/protobuf/proto"
+	ipfscore "github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+const migration012_ListingVersionForMarkupListings = 4
+
+type Migration012 struct{}
+
+type Migration012_ListingData struct {
+	Hash         string   `json:"hash"`
+	Slug         string   `json:"slug"`
+	Title        string   `json:"title"`
+	Categories   []string `json:"categories"`
+	NSFW         bool     `json:"nsfw"`
+	ContractType string   `json:"contractType"`
+	Description  string   `json:"description"`
+	Thumbnail    struct {
+		Tiny   string `json:"tiny"`
+		Small  string `json:"small"`
+		Medium string `json:"medium"`
+	} `json:"thumbnail"`
+	Price struct {
+		CurrencyCode string  `json:"currencyCode"`
+		Amount       uint64  `json:"amount"`
+		Modifier     float32 `json:"modifier"`
+	} `json:"price"`
+	ShipsTo            []string `json:"shipsTo"`
+	FreeShipping       []string `json:"freeShipping"`
+	Language           string   `json:"language"`
+	AverageRating      float32  `json:"averageRating"`
+	RatingCount        uint32   `json:"ratingCount"`
+	ModeratorIDs       []string `json:"moderators"`
+	AcceptedCurrencies []string `json:"acceptedCurrencies"`
+	CoinType           string   `json:"coinType"`
+}
+
+func Migration012_listingHasNewFeaturesAndOldVersion(sl *pb.SignedListing) bool {
+	return sl.Listing.Metadata.Version == 3 &&
+		sl.Listing.Metadata.PriceModifier != 0
```

Comment:
Is there any chance that any part of this chain is nil and this panics?

Reply:
Not really, all listings should have a metadata, but since there's no static assurance I added a check anyways.